lib/src/repo.rs

```diff
@@ -1220,6 +1234,42 @@ where
         IdIndex(vec)
     }
 
+    /// Looks up entries with the given prefix, and collects values if matched
+    /// entries have unambiguous keys.
+    pub fn resolve_prefix_with<U>(
+        &self,
+        prefix: &HexPrefix,
+        mut value_mapper: impl FnMut(&V) -> U,
+    ) -> PrefixResolution<Vec<U>> {
+        let mut range = self.resolve_prefix_range(prefix).peekable();
+        if let Some((first_key, _)) = range.peek().copied() {
+            let maybe_entries: Option<Vec<_>> = range
+                .map(|(k, v)| (k == first_key).then(|| value_mapper(v)))
+                .collect();
+            if let Some(entries) = maybe_entries {
+                PrefixResolution::SingleMatch(entries)
+            } else {
+                PrefixResolution::AmbiguousMatch
+            }
+        } else {
+            PrefixResolution::NoMatch
+        }
+    }
+
+    /// Iterates over entries with the given prefix.
+    pub fn resolve_prefix_range<'a: 'b, 'b>(
+        &'a self,
+        prefix: &'b HexPrefix,
+    ) -> impl Iterator<Item = (&'a K, &'a V)> + 'b {
+        let (bytes_prefix, min_bytes_prefix) = prefix.bytes_prefixes::<K>();
```

Comment:
Optional: I find this a bit hard to follow. It would help if it was documented what properties `bytes_prefix` and `min_bytes_prefix` satisfy. It would help even more if all of this optimization was hidden inside the implementation of HexPrefix.

The two suggestions below are the first things that came to mind, though I'm not sure if either is actually doable / a good idea.

It's fine to leave this as is, an think about whether this can be made clearer later.

Reply:
I'll add doc comment to `HexPrefix` in another PR.